opt: fix handling of write-only columns during update

andy-kimball opened the pull request:
A column in the write-only state has either been recently added or dropped
from a table. If recently added, then it's currently being backfilled, and
therefore may still have a NULL value. If NULL, then it's not valid to use
this value during an UPDATE or UPSERT. However, this is what the previous
code was doing; when a column value was updated, it would mistakenly
incorporate the NULL value into the update.

The fix is to *always* write the default or computed value of a column
during an UPDATE or UPSERT of a write-only column. Here are the rules:

  1. If column is computed, evaluate that expression as its value.
  2. If column has a default value specified for it, use that as its value.
  3. If column is nullable, use NULL as its value.
  4. If column is currently being added or dropped (i.e. a mutation column),
     use a default value (0 for INT column, "" for STRING column, etc).

One side effect of doing this is that NOT NULL columns can now be dropped
without triggering null constraint violations when other transactions try
to insert rows during the dropping process. Also, other transactions can
observe the default values if they SELECT a column that is dropped at just
the right time.

Fixes #42459

Release justification: Fix for high-severity bug in existing functionality.
This bug can result in NULL values being written to NOT NULL columns. It
can also trigger unexpected errors during INSERT/UPDATE/UPSERT statements
when schema changes are taking place.

Release note (sql change): Columns in the process of being added or removed
to a table are now always set to their default or computed value if another
transaction concurrently inserts, updates, or upserts a row. This fixes an
issue where a column being backfilled would not get properly set by
concurrent transactions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46285)
<!-- Reviewable:end -->

andy-kimball commented:
This is mostly an optimizer change, but thought that others who commented on the original issue might want to review the `NewDefaultDatum` method that provides default values for NOT NULL mutation columns.

RaduBerinde reviewed (commented):
Nice work! Will make another pass over it, but looks great so far!

[nit] can you add some information about the new trading queries in the commit message (or separate them in a second commit with a description)? It can just be a copy of the top-level file comment you have already. It will be nice to have a bit of information of them in the git history.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-:-M2npbtc9ciA7QerQ3Qo:bz4j5rl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @jordanlewis, @knz, and @RaduBerinde)

___
*[pkg/sql/opt/exec/execbuilder/testdata/schema_change_in_txn, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2npMWF8l7Z7Z4a2fUZ:-M2npMWF8l7Z7Z4a2fU_:b-r99qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/exec/execbuilder/testdata/schema_change_in_txn#L26)):*
> ```
> ALTER TABLE t ADD COLUMN g INT NOT NULL DEFAULT(15)
> 
> query T kvtrace(prefix=/Table/53/)
> ```

[nit] add a comment explaining that we expect default values for b,c, NULL value for e, and zero value for d.
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2noFmY8mu5c1iYIh1u:-M2noFmY8mu5c1iYIh1v:bfk8k0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/optbuilder/testdata/update#L1329)):*
> ```
> # ------------------------------------------------------------------------------
> 
> # Test update that doesn't require mutation column to be recalculated.
> ```

This comment doesn't look right anymore (and the one below)
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 1175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2noiY8AXj2yqsuTJ58:-M2noiY9-E-yYe6G9CVn:bec1qpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/optbuilder/testdata/upsert#L1175)):*
> ```
>            └── upsert_a:13 > 0 [as=check2:15]
> 
> # Don't directly update mutation columns. However, computed columns do need to
> ```

Is this comment still accurate?
___
*[pkg/sql/sem/tree/datum.go, line 3901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2nnDES6ntM-QJlbOBs:-M2nnDES6ntM-QJlbOBt:b-jb0jg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/sem/tree/datum.go#L3901)):*
> ```Go
> 		return bitArrayZero, nil
> 	default:
> 		panic(fmt.Sprintf("unhandled type %v", t.SQLString()))
> ```

[nit] should this be an error given that we can return one


<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: fix handling of write-only columns during update

A column in the write-only state has either been recently added or dropped
from a table. If recently added, then it's currently being backfilled, and
therefore may still have a NULL value. If NULL, then it's not valid to use
this value during an UPDATE or UPSERT. However, this is what the previous
code was doing; when a column value was updated, it would mistakenly
incorporate the NULL value into the update.

The fix is to *always* write the default or computed value of a column
during an UPDATE or UPSERT of a write-only column. Here are the rules:

  1. If column is computed, evaluate that expression as its value.
  2. If column has a default value specified for it, use that as its value.
  3. If column is nullable, use NULL as its value.
  4. If column is currently being added or dropped (i.e. a mutation column),
     use a default value (0 for INT column, "" for STRING column, etc).

One side effect of doing this is that NOT NULL columns can now be dropped
without triggering null constraint violations when other transactions try
to insert rows during the dropping process. Also, other transactions can
observe the default values if they SELECT a column that is dropped at just
the right time.

Fixes #42459

Release justification: Fix for high-severity bug in existing functionality.
This bug can result in NULL values being written to NOT NULL columns. It
can also trigger unexpected errors during INSERT/UPDATE/UPSERT statements
when schema changes are taking place.

Release note (sql change): Columns in the process of being added or removed
to a table are now always set to their default or computed value if another
transaction concurrently inserts, updates, or upserts a row. This fixes an
issue where a column being backfilled would not get properly set by
concurrent transactions.

andy-kimball committed:
opt: add new external test suite

Add a new external test suite that is based on a business buying/selling
online trading cards. There are two files. One lists the SELECT and mutation
query plans when there are no schema changes running concurrently. The other
shows how the query plans change when there are one or more column adds/drops
running concurrently. The purpose of this is to ensure that in-progress
schema changes do not materially change the plans.

Release justification: Non-production code change.
Release note: None

andy-kimball reviewed (commented):
I split the PR into two commits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-:-M2oM_XQ6iT16HstaDIG:b-i4a65j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @knz)

___
*[pkg/sql/opt/exec/execbuilder/testdata/schema_change_in_txn, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2npMWF8l7Z7Z4a2fUZ:-M2oMxWrC-WCoVnqiTLX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/exec/execbuilder/testdata/schema_change_in_txn#L26)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment explaining that we expect default values for b,c, NULL value for e, and zero value for d.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2noFmY8mu5c1iYIh1u:-M2oNRGcApWJt9K7Tlhm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/optbuilder/testdata/update#L1329)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This comment doesn't look right anymore (and the one below)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 1175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2noiY8AXj2yqsuTJ58:-M2oNjz-08bYS-wrHbaf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/opt/optbuilder/testdata/upsert#L1175)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Is this comment still accurate?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/datum.go, line 3901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2nnDES6ntM-QJlbOBs:-M2oNvTm2-SD0ws6Yi0P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/sem/tree/datum.go#L3901)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] should this be an error given that we can return one
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
The default datum values <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Didn't look at anything else. 

> Reviewed 1 of 14 files at r1, 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-:-M2oQhzk2auRx1BpJIk8:b-79pcgf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @knz)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

We should add some comments before the `buildScan` calls in `buildInputForUpsert` and `buildInputForUpdate` explaining why we are reading mutation values, and pointing out that we aren't (and shouldn't be) using them to calculate any "new" values.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-:-M2oneyA1IA3PTZeI3uY:bcu7r20)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, and @knz)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 509 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2on9pb8z6Kr4aMgdsT:-M2on9pc4VyHG_4TieXS:b6k5p4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6662d870461bde0d39528ef658e8a2e954d644/pkg/sql/opt/optbuilder/mutation_builder.go#L509)):*
> ```Go
> //   4. If column is currently being added or dropped (i.e. a mutation column),
> //      use a default value (0 for INT column, "" for STRING column, etc).
> //
> ```

[nit] add an explanation of why it's not ok to use the fetched values (as we've been doing so far).
___
*[pkg/sql/opt/xform/testdata/external/trading, line 544 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2oou8O99wyZlf0WNGM:-M2oou8P7al9aI9tMZ8x:b-nec2kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/18acc88324f4b20fa8f90d39397b209b65272cfd/pkg/sql/opt/xform/testdata/external/trading#L544)):*
> ```
> # Find all cards that have been modified in the last 5 seconds.
> #
> # Problems:
> ```

These queries and analysis are very cool!
___
*[pkg/sql/opt/xform/testdata/external/trading-mutation, line 756 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2ooaSDEHCo2gySFzW5:-M2ooaSDEHCo2gySFzW6:b-teoq0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/18acc88324f4b20fa8f90d39397b209b65272cfd/pkg/sql/opt/xform/testdata/external/trading-mutation#L756)):*
> ```
> #
> # Problems:
> #   1. This query will not compile, due to #46180.
> ```

This should be fixed now.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
I added some comments to `buildScan`. We actually need those values for the `Delete` case as well. The execution engine panics if they're not available, even for `Delete`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-:-M2p9StNEYo1T6UtF99b:b185ybt)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @jordanlewis and @knz)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 509 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2on9pb8z6Kr4aMgdsT:-M2pA7rSDjH402AsBcTm:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6662d870461bde0d39528ef658e8a2e954d644/pkg/sql/opt/optbuilder/mutation_builder.go#L509)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add an explanation of why it's not ok to use the fetched values (as we've been doing so far).
</blockquote></details>

Done
___
*[pkg/sql/opt/xform/testdata/external/trading-mutation, line 756 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2ooaSDEHCo2gySFzW5:-M2pF_eP26lA70rq8AGC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18acc88324f4b20fa8f90d39397b209b65272cfd/pkg/sql/opt/xform/testdata/external/trading-mutation#L756)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should be fixed now.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

rytaft commented:
___
*[pkg/sql/descriptor_mutation_test.go, line 806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46285#-M2o90wzCCLTq4BWar2b:-M2o90x--ES8WnngMHo7:b-7kpogb) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e8eac82b0c05fef31145e6c6f195559c879dde/pkg/sql/descriptor_mutation_test.go#L806)):*
> ```Go
> 
> 				if idxState == sqlbase.DescriptorMutation_DELETE_ONLY {
> 					// Index entry for row "a" is deleted.
> ```

Why did you change this to "a" (here and below)?

(Sorry I forgot to publish this earlier)


<!-- Sent from Reviewable.io -->
